host_source.go

```diff
@@ -271,106 +365,279 @@ func checkSystemLocal(control *controlConn) (bool, error) {
+				return fmt.Errorf(assertErrorMsg, "tokens"), nil
+			}
+		case "dse_version":
+			host.dseVersion, ok = value.(string)
+			if !ok {
+				return fmt.Errorf(assertErrorMsg, "dse_version"), nil
 			}
 		}
+		// TODO(thrawn01): Add 'port'? once CASSANDRA-7544 is complete
+		// Not sure what the port field will be called until the JIRA issue is complete
+	}
 
-		return false, err
+	// Default to our connected port if the cluster doesn't have port information
+	if host.port == 0 {
+		host.port = r.session.cfg.Port
 	}
 
-	return true, nil
+	return nil, &host
 }
 
-func (r *ringDescriber) GetHosts() (hosts []*HostInfo, partitioner string, err error) {
-	r.mu.Lock()
-	defer r.mu.Unlock()
-	// we need conn to be the same because we need to query system.peers and system.local
-	// on the same node to get the whole cluster
-
-	const (
-		legacyLocalQuery = "SELECT data_center, rack, host_id, tokens, partitioner, release_version FROM system.local"
-		// only supported in 2.2.0, 2.1.6, 2.0.16
-		localQuery = "SELECT broadcast_address, data_center, rack, host_id, tokens, partitioner, release_version FROM system.local"
-	)
-
-	localHost := &HostInfo{}
-	if r.localHasRpcAddr {
-		iter := r.session.control.query(localQuery)
-		if iter == nil {
-			return r.prevHosts, r.prevPartitioner, nil
-		}
+// Ask the control node for it's local host information
+func (r *ringDescriber) GetLocalHostInfo() (*HostInfo, error) {
+	row := make(map[string]interface{})
 
-		iter.Scan(&localHost.peer, &localHost.dataCenter, &localHost.rack,
-			&localHost.hostId, &localHost.tokens, &partitioner, &localHost.version)
+	// ask the connected node for local host info
+	it := r.session.control.query("SELECT * FROM system.local WHERE key='local'")
+	if it == nil {
+		return nil, errors.New("Attempted to query 'system.local' on a closed control connection")
+	}
 
-		if err = iter.Close(); err != nil {
-			return nil, "", err
-		}
-	} else {
-		iter := r.session.control.withConn(func(c *Conn) *Iter {
-			localHost = c.host
-			return c.query(legacyLocalQuery)
-		})
+	// expect only 1 row
+	it.MapScan(row)
+	if err := it.Close(); err != nil {
+		return nil, err
+	}
 
-		if iter == nil {
-			return r.prevHosts, r.prevPartitioner, nil
-		}
+	// extract all available info about the host
+	err, host := r.hostInfoFromMap(row)
+	if err != nil {
+		return nil, err
+	}
+
+	return host, err
+}
 
-		iter.Scan(&localHost.dataCenter, &localHost.rack, &localHost.hostId, &localHost.tokens, &partitioner, &localHost.version)
+// Given an ip address and port, return a peer that matched the ip address
+func (r *ringDescriber) GetPeerHostInfo(ip net.IP, port int) (*HostInfo, error) {
+	row := make(map[string]interface{})
 
-		if err = iter.Close(); err != nil {
-			return nil, "", err
-		}
+	it := r.session.control.query("SELECT * FROM system.peers WHERE peer=?", ip)
+	if it == nil {
+		return nil, errors.New("Attempted to query 'system.peers' on a closed control connection")
+	}
+
+	// expect only 1 row
```

Comment:
these functions all share common code, maybe split the common code out to something like

```go
func (r *ringDescriber) extractHostInfo(it Iter) (*HostInfo, error) {
...
}
```